In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import mean_squared_error
%matplotlib inline

In [2]:
df = pd.read_csv('Dataset/train.csv')

In [3]:
df_test = pd.read_csv('Dataset/test.csv')

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550068 entries, 0 to 550067
Data columns (total 12 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   User_ID                     550068 non-null  int64  
 1   Product_ID                  550068 non-null  object 
 2   Gender                      550068 non-null  object 
 3   Age                         550068 non-null  object 
 4   Occupation                  550068 non-null  int64  
 5   City_Category               550068 non-null  object 
 6   Stay_In_Current_City_Years  550068 non-null  object 
 7   Marital_Status              550068 non-null  int64  
 8   Product_Category_1          550068 non-null  int64  
 9   Product_Category_2          376430 non-null  float64
 10  Product_Category_3          166821 non-null  float64
 11  Purchase                    550068 non-null  int64  
dtypes: float64(2), int64(5), object(5)
memory usage: 50.4+ MB


In [5]:
df_test.head()

User_ID Product_ID Gender    Age  Occupation City_Category  \
0  1000004  P00128942      M  46-50           7             B   
1  1000009  P00113442      M  26-35          17             C   
2  1000010  P00288442      F  36-45           1             B   
3  1000010  P00145342      F  36-45           1             B   
4  1000011  P00053842      F  26-35           1             C   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               1                   1   
1                          0               0                   3   
2                         4+               1                   5   
3                         4+               1                   4   
4                          1               0                   4   

   Product_Category_2  Product_Category_3  
0                11.0                 NaN  
1                 5.0                 NaN  
2                14.0                 NaN  
3                 9.0                 NaN  
4                 5.0                12.0

In [6]:
df.head()

User_ID Product_ID Gender   Age  Occupation City_Category  \
0  1000001  P00069042      F  0-17          10             A   
1  1000001  P00248942      F  0-17          10             A   
2  1000001  P00087842      F  0-17          10             A   
3  1000001  P00085442      F  0-17          10             A   
4  1000002  P00285442      M   55+          16             C   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               0                   3   
1                          2               0                   1   
2                          2               0                  12   
3                          2               0                  12   
4                         4+               0                   8   

   Product_Category_2  Product_Category_3  Purchase  
0                 NaN                 NaN      8370  
1                 6.0                14.0     15200  
2                 NaN                 NaN      1422  
3                14.0                 NaN      1057  
4                 NaN                 NaN      7969

In [7]:
train_df = df.iloc[:,:-1]
train_df.head()

User_ID Product_ID Gender   Age  Occupation City_Category  \
0  1000001  P00069042      F  0-17          10             A   
1  1000001  P00248942      F  0-17          10             A   
2  1000001  P00087842      F  0-17          10             A   
3  1000001  P00085442      F  0-17          10             A   
4  1000002  P00285442      M   55+          16             C   

  Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                          2               0                   3   
1                          2               0                   1   
2                          2               0                  12   
3                          2               0                  12   
4                         4+               0                   8   

   Product_Category_2  Product_Category_3  
0                 NaN                 NaN  
1                 6.0                14.0  
2                 NaN                 NaN  
3                14.0                 NaN  
4                 NaN                 NaN

In [8]:
train_df['check'] = 1
df_test['check'] = 0

In [9]:
train_df.shape, df_test.shape

((550068, 12), (233599, 12))

In [10]:
#concate the train set and test set to create 
df_concat = pd.concat([train_df, df_test], ignore_index=True)

In [11]:
df_concat

User_ID Product_ID Gender    Age  Occupation City_Category  \
0       1000001  P00069042      F   0-17          10             A   
1       1000001  P00248942      F   0-17          10             A   
2       1000001  P00087842      F   0-17          10             A   
3       1000001  P00085442      F   0-17          10             A   
4       1000002  P00285442      M    55+          16             C   
...         ...        ...    ...    ...         ...           ...   
783662  1006036  P00118942      F  26-35          15             B   
783663  1006036  P00254642      F  26-35          15             B   
783664  1006036  P00031842      F  26-35          15             B   
783665  1006037  P00124742      F  46-50           1             C   
783666  1006039  P00316642      F  46-50           0             B   

       Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                               2               0                   3   
1                               2               0                   1   
2                               2               0                  12   
3                               2               0                  12   
4                              4+               0                   8   
...                           ...             ...                 ...   
783662                         4+               1                   8   
783663                         4+               1                   5   
783664                         4+               1                   1   
783665                         4+               0                  10   
783666                         4+               1                   4   

        Product_Category_2  Product_Category_3  check  
0                      NaN                 NaN      1  
1                      6.0                14.0      1  
2                      NaN                 NaN      1  
3                     14.0                 NaN      1  
4                      NaN                 NaN      1  
...                    ...                 ...    ...  
783662                 NaN                 NaN      0  
783663                 8.0                 NaN      0  
783664                 5.0                12.0      0  
783665                16.0                 NaN      0  
783666                 5.0                 NaN      0  

[783667 rows x 12 columns]

In [12]:
df_concat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 783667 entries, 0 to 783666
Data columns (total 12 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   User_ID                     783667 non-null  int64  
 1   Product_ID                  783667 non-null  object 
 2   Gender                      783667 non-null  object 
 3   Age                         783667 non-null  object 
 4   Occupation                  783667 non-null  int64  
 5   City_Category               783667 non-null  object 
 6   Stay_In_Current_City_Years  783667 non-null  object 
 7   Marital_Status              783667 non-null  int64  
 8   Product_Category_1          783667 non-null  int64  
 9   Product_Category_2          537685 non-null  float64
 10  Product_Category_3          237858 non-null  float64
 11  check                       783667 non-null  int64  
dtypes: float64(2), int64(5), object(5)
memory usage: 71.7+ MB


In [13]:
df_concat.fillna(-999, inplace=True)

In [14]:
df_concat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 783667 entries, 0 to 783666
Data columns (total 12 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   User_ID                     783667 non-null  int64  
 1   Product_ID                  783667 non-null  object 
 2   Gender                      783667 non-null  object 
 3   Age                         783667 non-null  object 
 4   Occupation                  783667 non-null  int64  
 5   City_Category               783667 non-null  object 
 6   Stay_In_Current_City_Years  783667 non-null  object 
 7   Marital_Status              783667 non-null  int64  
 8   Product_Category_1          783667 non-null  int64  
 9   Product_Category_2          783667 non-null  float64
 10  Product_Category_3          783667 non-null  float64
 11  check                       783667 non-null  int64  
dtypes: float64(2), int64(5), object(5)
memory usage: 71.7+ MB


In [15]:
df_concat['Product_Category_3'].unique()

array([-999.,   14.,   17.,    5.,    4.,   16.,   15.,    8.,    9.,
         13.,    6.,   12.,    3.,   18.,   11.,   10.])

In [16]:
df_concat['User_ID'].describe()

count    7.836670e+05
mean     1.003029e+06
std      1.727267e+03
min      1.000001e+06
25%      1.001519e+06
50%      1.003075e+06
75%      1.004478e+06
max      1.006040e+06
Name: User_ID, dtype: float64

In [17]:
print(df_concat['Gender'].unique())

['F' 'M']


In [18]:
print(sorted(df_concat['Occupation'].unique()))
df_concat.groupby('Occupation')['Occupation'].count()

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]


Occupation
0      98850
1      67687
2      37996
3      25126
4     103086
5      17373
6      29102
7      84127
8       2189
9       8929
10     18304
11     16593
12     44448
13     11053
14     38782
15     17356
16     36122
17     57418
18      9367
19     11919
20     47840
Name: Occupation, dtype: int64

In [19]:
print(df_concat['Age'].unique())
df_concat['Age'].describe()

['0-17' '55+' '26-35' '46-50' '51-55' '36-45' '18-25']


count     783667
unique         7
top        26-35
freq      313015
Name: Age, dtype: object

In [20]:
print(df_concat['City_Category'].unique())
print(df_concat['City_Category'].describe())

['A' 'C' 'B']
count     783667
unique         3
top            B
freq      329739
Name: City_Category, dtype: object


In [21]:
print(df_concat['Stay_In_Current_City_Years'].unique())
print(df_concat['Stay_In_Current_City_Years'].describe())
print(df_concat.groupby('Stay_In_Current_City_Years')['Stay_In_Current_City_Years'].count())

['2' '4+' '3' '1' '0']
count     783667
unique         5
top            1
freq      276425
Name: Stay_In_Current_City_Years, dtype: object
Stay_In_Current_City_Years
0     105716
1     276425
2     145427
3     135428
4+    120671
Name: Stay_In_Current_City_Years, dtype: int64


In [22]:
print(df_concat['Marital_Status'].unique())
print(df_concat['Marital_Status'].describe())
print(df_concat.groupby('Marital_Status')['Marital_Status'].count())

[0 1]
count    783667.000000
mean          0.409777
std           0.491793
min           0.000000
25%           0.000000
50%           0.000000
75%           1.000000
max           1.000000
Name: Marital_Status, dtype: float64
Marital_Status
0    462538
1    321129
Name: Marital_Status, dtype: int64


In [23]:
print(df_concat['Product_Category_1'].unique())
print(df_concat['Product_Category_1'].describe())
print(df_concat.groupby('Product_Category_1')['Product_Category_1'].count())

[ 3  1 12  8  5  4  2  6 14 11 13 15  7 16 18 10 17  9 20 19]
count    783667.000000
mean          5.366196
std           3.878160
min           1.000000
25%           1.000000
50%           5.000000
75%           8.000000
max          20.000000
Name: Product_Category_1, dtype: float64
Product_Category_1
1     200699
2      34056
3      28791
4      16756
5     215950
6      29326
7       5345
8     162294
9        604
10      7373
11     34440
12      5610
13      7930
14      2186
15      8984
16     13933
17       801
18      4436
19      1603
20      2550
Name: Product_Category_1, dtype: int64


In [24]:
print(sorted(df_concat['Product_Category_2'].unique()))
print(df_concat['Product_Category_2'].describe())
print(df_concat.groupby('Product_Category_2')['Product_Category_2'].count())
df_concat['Product_Category_2'].isnull().sum()

[-999.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0]
count    783667.000000
mean       -306.817532
std         468.194361
min        -999.000000
25%        -999.000000
50%           5.000000
75%          14.000000
max          18.000000
Name: Product_Category_2, dtype: float64
Product_Category_2
-999.0    245982
 2.0       70498
 3.0        4123
 4.0       36705
 5.0       37165
 6.0       23575
 7.0         854
 8.0       91317
 9.0        8177
 10.0       4420
 11.0      20230
 12.0       7801
 13.0      15054
 14.0      78834
 15.0      54114
 16.0      61687
 17.0      19104
 18.0       4027
Name: Product_Category_2, dtype: int64


0

In [25]:
print(sorted(df_concat['Product_Category_3'].unique()))
print(df_concat['Product_Category_3'].describe())
print(df_concat.groupby('Product_Category_3')['Product_Category_3'].count())
df_concat['Product_Category_3'].isnull().sum()

[-999.0, 3.0, 4.0, 5.0, 6.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0]
count    783667.000000
mean       -691.939130
std         465.148262
min        -999.000000
25%        -999.000000
50%        -999.000000
75%           8.000000
max          18.000000
Name: Product_Category_3, dtype: float64
Product_Category_3
-999.0    545809
 3.0         878
 4.0        2691
 5.0       23799
 6.0        6888
 8.0       17861
 9.0       16532
 10.0       2501
 11.0       2585
 12.0      13115
 13.0       7849
 14.0      26283
 15.0      39968
 16.0      46469
 17.0      23818
 18.0       6621
Name: Product_Category_3, dtype: int64


0

## Initial Analysis of features and target
* Gender has two categorical values F and M. So we will have to convert Gender from Object data type to categorical
* Occupation has 21 unique categories from 0 to 21. So it should be converted to categorical.
* Age is divided into 7 unique intervals. So convert it into categorical.
* Three categories of citys are there. So convert to categorical.
* Stay_In_Current_City_Years has 5 unique values. So convert to categorical.
* Marital Status has two values 0 and 1. So convert to categorical.
* Product_Category_1 has 20 unique values. Convert to categorical.
* Product_Category_2 had 173638 values missing specified as NaN. These converted to -999 value.
* Product_Category_3 has 383247 values missing specified as NaN. These converted to -999 value.
* Purchase is the target variable. None are missing.

Convert Product_ID and Customer_ID to categorical as well.

In [26]:
df_concat_copy = df_concat.copy()

New product category 'final_category' has a total of 237 category types which could be converted to categorical values.

In [27]:
user_id_encoder = LabelEncoder()
df_concat['User_ID'] = user_id_encoder.fit_transform(df_concat['User_ID'])

In [28]:
product_id_encoder = LabelEncoder()
df_concat['Product_ID'] = product_id_encoder.fit_transform(df_concat['Product_ID'])
age_encoder = LabelEncoder()
df_concat['Age'] = age_encoder.fit_transform(df_concat['Age'])
occupation_encoder = LabelEncoder()
df_concat['Occupation'] = occupation_encoder.fit_transform(df_concat['Occupation'])
city_category_encoder = LabelEncoder()
df_concat['City_Category'] = city_category_encoder.fit_transform(df_concat['City_Category'])
stay_encoder = LabelEncoder()
df_concat['Stay_In_Current_City_Years'] = stay_encoder.fit_transform(df_concat['Stay_In_Current_City_Years'])

In [29]:
gender_encoder = LabelEncoder()
df_concat['Gender'] = gender_encoder.fit_transform(df_concat['Gender'])

In [30]:
df_concat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 783667 entries, 0 to 783666
Data columns (total 12 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   User_ID                     783667 non-null  int64  
 1   Product_ID                  783667 non-null  int32  
 2   Gender                      783667 non-null  int32  
 3   Age                         783667 non-null  int32  
 4   Occupation                  783667 non-null  int64  
 5   City_Category               783667 non-null  int32  
 6   Stay_In_Current_City_Years  783667 non-null  int32  
 7   Marital_Status              783667 non-null  int64  
 8   Product_Category_1          783667 non-null  int64  
 9   Product_Category_2          783667 non-null  float64
 10  Product_Category_3          783667 non-null  float64
 11  check                       783667 non-null  int64  
dtypes: float64(2), int32(5), int64(5)
memory usage: 56.8 MB


In [31]:
df_concat.head()

User_ID  Product_ID  Gender  Age  Occupation  City_Category  \
0        0         684       0    0          10              0   
1        0        2406       0    0          10              0   
2        0         868       0    0          10              0   
3        0         844       0    0          10              0   
4        1        2769       1    6          16              2   

   Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
0                           2               0                   3   
1                           2               0                   1   
2                           2               0                  12   
3                           2               0                  12   
4                           4               0                   8   

   Product_Category_2  Product_Category_3  check  
0              -999.0              -999.0      1  
1                 6.0                14.0      1  
2              -999.0              -999.0      1  
3                14.0              -999.0      1  
4              -999.0              -999.0      1

In [32]:
X = df_concat.iloc[:550068,:]
y = df.iloc[:550068,-1:]

In [33]:
X = X.drop('check', axis='columns')

In [34]:
X.shape, y.shape

((550068, 11), (550068, 1))

In [35]:
X.tail()

User_ID  Product_ID  Gender  Age  Occupation  City_Category  \
550063     5883        3612       1    5          13              1   
550064     5885        3613       0    2           1              2   
550065     5886        3613       0    2          15              1   
550066     5888        3613       0    6           1              2   
550067     5889        3611       0    4           0              1   

        Stay_In_Current_City_Years  Marital_Status  Product_Category_1  \
550063                           1               1                  20   
550064                           3               0                  20   
550065                           4               1                  20   
550066                           2               0                  20   
550067                           4               1                  20   

        Product_Category_2  Product_Category_3  
550063              -999.0              -999.0  
550064              -999.0              -999.0  
550065              -999.0              -999.0  
550066              -999.0              -999.0  
550067              -999.0              -999.0

In [36]:
y.tail()

Purchase
550063       368
550064       371
550065       137
550066       365
550067       490

In [37]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [38]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((385047, 11), (165021, 11), (385047, 1), (165021, 1))

In [39]:
y_test.head()

Purchase
84432      19142
72724      15513
197032      7802
353704     15455
91198       4492

In [40]:
from sklearn.tree import DecisionTreeRegressor

In [41]:
tree_reg = DecisionTreeRegressor(min_samples_split=8, min_samples_leaf = 80)

In [42]:
tree_reg.fit(X_train, y_train)

DecisionTreeRegressor(min_samples_leaf=80, min_samples_split=8)

In [43]:
print(tree_reg.feature_importances_)

[4.19834149e-03 7.81515824e-02 9.51609602e-04 2.17042828e-03
 1.84677138e-03 3.03043218e-03 8.56554758e-04 4.64201591e-04
 8.81308440e-01 1.48428704e-02 1.21787679e-02]


In [44]:
predictions = tree_reg.predict(X_train)

In [45]:
test_predictions = tree_reg.predict(X_test)

In [46]:
np.sqrt(mean_squared_error(y_train, predictions))

2638.805466210935

In [47]:
np.sqrt(mean_squared_error(y_test, test_predictions))

2726.056848357811

After doing some experimentations with decision tree the best validation score turn out to be 2726 when model developed
min_samples_split=8 and min_samples_leaf=80.

Now experimenting with Linear Model.

In [48]:
from sklearn.linear_model import LinearRegression

In [49]:
simple_linear = LinearRegression()
simple_linear.fit(X_train, y_train)

LinearRegression()

In [50]:
train_predicts = simple_linear.predict(X_train)
test_predicts = simple_linear.predict(X_test)
print(np.sqrt(mean_squared_error(y_train, train_predicts)))
print(np.sqrt(mean_squared_error(y_test, test_predicts)))

4600.104780342809
4602.7333161336965


As can be seen above even the training error is very high : 4600 without any regularization. Thus linear model clearly
underfits for this purpose.

Now lets try Random Forest.

In [51]:
from sklearn.ensemble import RandomForestRegressor

In [65]:
forest_regressor = RandomForestRegressor(min_samples_split=8, min_samples_leaf = 80, n_estimators=30)

In [66]:
forest_regressor.fit(X_train, y_train)

<ipython-input-66-f6e921078831>:1: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  forest_regressor.fit(X_train, y_train)


RandomForestRegressor(min_samples_leaf=80, min_samples_split=80,
                      n_estimators=30)

In [54]:
train_predicts = forest_regressor.predict(X_train)
test_predicts = forest_regressor.predict(X_test)
print(np.sqrt(mean_squared_error(y_train, train_predicts)))
print(np.sqrt(mean_squared_error(y_test, test_predicts)))

2645.1705915631683
2710.6363828600233


Now running on test set provided

In [55]:
test_data = df_concat.loc[550068:,:]

In [56]:
test_data = test_data.drop('check', axis='columns')

In [57]:
predictions = forest_regressor.predict(test_data)

In [58]:
preds_df = pd.DataFrame(predictions, columns=['Purchase'])

In [59]:
preds_df.head()

Purchase
0  15674.179643
1  11649.705179
2   6247.861389
3   2594.372100
4   2548.911334

In [60]:
predictions

array([15674.17964323, 11649.70517941,  6247.86138868, ...,
       11291.78651692, 18554.77418184,  2662.6396505 ])

In [61]:
output = df_test.loc[:,['User_ID', 'Product_ID']]

In [62]:
output = pd.concat([output,preds_df], axis=1)

In [63]:
output.tail()

User_ID Product_ID      Purchase
233594  1006036  P00118942   7928.638954
233595  1006036  P00254642   6525.563328
233596  1006036  P00031842  11291.786517
233597  1006037  P00124742  18554.774182
233598  1006039  P00316642   2662.639650

In [64]:
output.to_csv('dataset/Submission1.csv', index=False)

As can be seen above Random Forest is performing a little bit better as compared to Decision Tree alone.

The score for test case for this random forest is 2712.6847733051.